src/Stream/BufferedStream.php

```diff
@@ -0,0 +1,264 @@
+<?php
+
+namespace Http\Message\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A buffered stream allow to buffer over an existing.
+ *
+ * You should use this decorator when you want to seek over a not seekable stream.
+ * This stream is however read-only.
```

Comment:
Please correct me if Im wrong but I do not think it would ever lead to a memory problem. We use max 2MB memory as default (see constructor)